File: appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroBaseFragment.java
package com.github.paolorotolo.appintro;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.github.paolorotolo.appintro.internal.LogHelper;
import com.github.paolorotolo.appintro.internal.TypefaceContainer;

import androidx.annotation.ColorInt;
import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

public abstract class AppIntroBaseFragment extends Fragment implements ISlideSelectionListener,
        ISlideBackgroundColorHolder {
    protected static final String ARG_TITLE = "title";
    protected static final String ARG_TITLE_TYPEFACE = "title_typeface";
    protected static final String ARG_TITLE_TYPEFACE_RES = "title_typeface_res";
    protected static final String ARG_DESC = "desc";
    protected static final String ARG_DESC_TYPEFACE = "desc_typeface";
    protected static final String ARG_DESC_TYPEFACE_RES = "desc_typeface_res";
    protected static final String ARG_DRAWABLE = "drawable";
    protected static final String ARG_BG_COLOR = "bg_color";
    protected static final String ARG_TITLE_COLOR = "title_color";
    protected static final String ARG_DESC_COLOR = "desc_color";

    private static final String TAG = LogHelper.makeLogTag(AppIntroBaseFragment.class);

    private int drawable, bgColor, titleColor, descColor, layoutId;
    private String title, description;
    private TypefaceContainer titleTypeface, descTypeface;

    private LinearLayout mainLayout;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);

        if (getArguments() != null && getArguments().size() != 0) {
            String argsTitleTypeface = getArguments().getString(ARG_TITLE_TYPEFACE);
            String argsDescTypeface = getArguments().getString(ARG_DESC_TYPEFACE);

            int argsTitleTypefaceRes = getArguments().getInt(ARG_TITLE_TYPEFACE_RES);
            int argsDescTypefaceRes = getArguments().getInt(ARG_DESC_TYPEFACE_RES);

            drawable = getArguments().getInt(ARG_DRAWABLE);

            title = getArguments().getString(ARG_TITLE);
            description = getArguments().getString(ARG_DESC);
            titleTypeface = new TypefaceContainer(argsTitleTypeface, argsTitleTypefaceRes);
            descTypeface = new TypefaceContainer(argsDescTypeface, argsDescTypefaceRes);

            bgColor = getArguments().getInt(ARG_BG_COLOR);
            titleColor = getArguments().containsKey(ARG_TITLE_COLOR) ?
                    getArguments().getInt(ARG_TITLE_COLOR) : 0;
            descColor = getArguments().containsKey(ARG_DESC_COLOR) ?
                    getArguments().getInt(ARG_DESC_COLOR) : 0;
        }
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        if (savedInstanceState != null) {
            drawable = savedInstanceState.getInt(ARG_DRAWABLE);
            title = savedInstanceState.getString(ARG_TITLE);
            description = savedInstanceState.getString(ARG_DESC);

            titleTypeface = new TypefaceContainer(
                    savedInstanceState.getString(ARG_TITLE_TYPEFACE),
                    savedInstanceState.getInt(ARG_TITLE_TYPEFACE_RES, 0));
            descTypeface = new TypefaceContainer(
                    savedInstanceState.getString(ARG_DESC_TYPEFACE),
                    savedInstanceState.getInt(ARG_DESC_TYPEFACE_RES, 0));

            bgColor = savedInstanceState.getInt(ARG_BG_COLOR);
            titleColor = savedInstanceState.getInt(ARG_TITLE_COLOR);
            descColor = savedInstanceState.getInt(ARG_DESC_COLOR);
        }
    }


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(getLayoutId(), container, false);
        TextView titleText = view.findViewById(R.id.title);
        TextView descriptionText = view.findViewById(R.id.description);
        ImageView slideImage = view.findViewById(R.id.image);
        mainLayout = view.findViewById(R.id.main);

        titleText.setText(title);
        if (titleColor != 0) {
            titleText.setTextColor(titleColor);
        }
        titleTypeface.applyTo(titleText);
        titleTypeface.applyTo(descriptionText);
        descriptionText.setText(description);
        if (descColor != 0) {
            descriptionText.setTextColor(descColor);
        }
        slideImage.setImageResource(drawable);
        mainLayout.setBackgroundColor(bgColor);

        return view;
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putInt(ARG_DRAWABLE, drawable);
        outState.putString(ARG_TITLE, title);
        outState.putString(ARG_DESC, description);
        outState.putInt(ARG_BG_COLOR, bgColor);
        outState.putInt(ARG_TITLE_COLOR, titleColor);
        outState.putInt(ARG_DESC_COLOR, descColor);
        saveTypefacesInstanceState(outState);
        super.onSaveInstanceState(outState);
    }

    private void saveTypefacesInstanceState(Bundle outState) {
        if (titleTypeface != null){
            outState.putString(ARG_TITLE_TYPEFACE, titleTypeface.getTypeFaceUrl());
            outState.putInt(ARG_TITLE_TYPEFACE_RES, titleTypeface.getTypeFaceResource());
        }
        if (descTypeface != null){
            outState.putString(ARG_DESC_TYPEFACE, descTypeface.getTypeFaceUrl());
            outState.putInt(ARG_DESC_TYPEFACE_RES, descTypeface.getTypeFaceResource());
        }
    }

    @Override
    public void onSlideDeselected() {
        LogHelper.d(TAG, String.format("Slide %s has been deselected.", title));
    }

    @Override
    public void onSlideSelected() {
        LogHelper.d(TAG, String.format("Slide %s has been selected.", title));
    }

    @Override
    public int getDefaultBackgroundColor() {
        return bgColor;
    }

    @Override
    public void setBackgroundColor(@ColorInt int backgroundColor) {
        mainLayout.setBackgroundColor(backgroundColor);
    }

    @LayoutRes
    protected abstract int getLayoutId();
}

File: appintro/src/main/java/com/github/appintro/AppIntroBaseFragment.kt
package com.github.appintro

import android.graphics.drawable.Animatable
import android.os.Bundle
import android.text.method.ScrollingMovementMethod
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.annotation.LayoutRes
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import com.appintro.core.contracts.SlideBackgroundColorHolder
import com.appintro.core.contracts.SlideSelectionListener
import com.appintro.core.utils.LogHelper
import com.appintro.core.utils.TypefaceContainer

abstract class AppIntroBaseFragment :
    Fragment(),
    SlideSelectionListener,
    SlideBackgroundColorHolder {
    private val viewModel: AppIntroFragmentViewModel by viewModels()

    private val logTAG = LogHelper.makeLogTag(AppIntroBaseFragment::class.java)

    @get:LayoutRes
    protected abstract val layoutId: Int

    @ColorInt
    @Deprecated(
        "`defaultBackgroundColor` has been deprecated to support configuration changes",
        ReplaceWith("defaultBackgroundColorRes"),
    )
    final override var defaultBackgroundColor: Int = 0
        private set

    @ColorRes
    final override var defaultBackgroundColorRes: Int = 0
        private set

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        @Suppress("DEPRECATION")
        defaultBackgroundColor = viewModel.defaultBackgroundColor ?: 0
        defaultBackgroundColorRes = viewModel.defaultBackgroundColorRes ?: 0

        val args = arguments
        if (args != null && args.size() != 0) {
            viewModel.drawable = args.getInt(ARG_DRAWABLE)
            viewModel.title = args.getCharSequence(ARG_TITLE)
            viewModel.description = args.getCharSequence(ARG_DESC)

            if (args.containsKey(ARG_BG_DRAWABLE)) {
                viewModel.bgDrawable = args.getInt(ARG_BG_DRAWABLE)
            }

            viewModel.titleTypefaceUrl = args.getString(ARG_TITLE_TYPEFACE_URL)
            viewModel.titleTypefaceRes = args.getInt(ARG_TITLE_TYPEFACE_RES)

            viewModel.descTypefaceUrl = args.getString(ARG_DESC_TYPEFACE_URL)
            viewModel.descTypefaceRes = args.getInt(ARG_DESC_TYPEFACE_RES)

            viewModel.defaultBackgroundColor = args.getInt(ARG_BG_COLOR)
            viewModel.defaultBackgroundColorRes = args.getInt(ARG_BG_COLOR_RES)

            if (args.containsKey(ARG_TITLE_COLOR)) {
                viewModel.titleColor = args.getInt(ARG_TITLE_COLOR)
            }

            if (args.containsKey(ARG_TITLE_COLOR_RES)) {
                viewModel.titleColorRes = args.getInt(ARG_TITLE_COLOR_RES)
            }

            if (args.containsKey(ARG_DESC_COLOR)) {
                viewModel.descColor = args.getInt(ARG_DESC_COLOR)
            }

            if (args.containsKey(ARG_DESC_COLOR_RES)) {
                viewModel.descColorRes = args.getInt(ARG_DESC_COLOR_RES)
            }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View? {
        val view = inflater.inflate(layoutId, container, false)
        val titleText = view.findViewById<TextView>(R.id.title)
        val descriptionText = view.findViewById<TextView>(R.id.description)
        val slideImage = view.findViewById<ImageView>(R.id.image)
        val mainLayout = view.findViewById<ConstraintLayout>(R.id.main)

        titleText.text = viewModel.title
        descriptionText.text = viewModel.description
        val titleColorRes = viewModel.titleColorRes
        val titleColor = viewModel.titleColor

        if (titleColorRes != null) {
            titleText.setTextColor(ContextCompat.getColor(requireContext(), titleColorRes))
        } else if (titleColor != null) { // Fallback to deprecated static color
            titleText.setTextColor(titleColor)
        }

        val descColorRes = viewModel.descColorRes
        val descColor = viewModel.descColor

        if (descColorRes != null) {
            descriptionText.setTextColor(ContextCompat.getColor(requireContext(), descColorRes))
        } else if (descColor != null) { // Fallback to deprecated static color
            descriptionText.setTextColor(descColor)
        }

        TypefaceContainer(
            typeFaceUrl = viewModel.titleTypefaceUrl,
            typeFaceResource = viewModel.titleTypefaceRes ?: 0,
        ).applyTo(titleText)

        TypefaceContainer(
            typeFaceUrl = viewModel.descTypefaceUrl,
            typeFaceResource = viewModel.descTypefaceRes ?: 0,
        ).applyTo(descriptionText)

        viewModel.drawable?.let {
            slideImage.setImageResource(it)
        }

        val bgDrawable = viewModel.bgDrawable

        when {
            bgDrawable != null -> {
                mainLayout?.setBackgroundResource(bgDrawable)
            }
            defaultBackgroundColorRes != 0 -> {
                mainLayout?.setBackgroundColor(ContextCompat.getColor(requireContext(), defaultBackgroundColorRes))
            }
            else -> {
                @Suppress("DEPRECATION")
                mainLayout?.setBackgroundColor(defaultBackgroundColor)
            }
        }

        titleText.movementMethod = ScrollingMovementMethod()
        descriptionText.movementMethod = ScrollingMovementMethod()

        return view
    }

    override fun onResume() {
        super.onResume()

        view?.findViewById<ImageView>(R.id.image).let {
            if (it is Animatable) {
                it.start()
            }
        }
    }

    override fun onPause() {
        super.onPause()

        view?.findViewById<ImageView>(R.id.image).let {
            if (it is Animatable) {
                it.start()
            }
        }
    }

    override fun onSlideDeselected() {
        LogHelper.d(logTAG, "Slide ${viewModel.title} has been deselected.")
    }

    override fun onSlideSelected() {
        LogHelper.d(logTAG, "Slide ${viewModel.title} has been selected.")
    }

    override fun setBackgroundColor(
        @ColorInt backgroundColor: Int,
    ) {
        view?.findViewById<ConstraintLayout>(R.id.main)?.setBackgroundColor(backgroundColor)
    }
}
